beefy-gadget/src/lib.rs

```diff
@@ -52,74 +55,89 @@ pub fn beefy_peers_set_config() -> sc_network::config::NonDefaultSetConfig {
 	}
 }
 
+/// A convenience BEEFY client trait that defines all the type bounds a BEEFY client
+/// has to satisfy. Ideally that should actually be a trait alias. Unfortunately as
+/// of today, Rust does not allow a type alias to be used as a trait bound. Tracking
+/// issue is <https://github.com/rust-lang/rust/issues/41517>.
+pub(crate) trait Client<B, BE, P>:
+	BlockchainEvents<B> + HeaderBackend<B> + Finalizer<B, BE> + ProvideRuntimeApi<B> + Send + Sync
+where
+	B: Block,
+	BE: Backend<B>,
+	P: sp_core::Pair,
+	P::Public: AppPublic + Codec,
+	P::Signature: Clone + Codec + Debug + PartialEq + TryFrom<Vec<u8>>,
+{
+	// empty
+}
+
+impl<B, BE, P, T> Client<B, BE, P> for T
+where
+	B: Block,
+	BE: Backend<B>,
+	P: sp_core::Pair,
+	P::Public: AppPublic + Codec,
+	P::Signature: Clone + Codec + Debug + PartialEq + TryFrom<Vec<u8>>,
+	T: BlockchainEvents<B> + HeaderBackend<B> + Finalizer<B, BE> + ProvideRuntimeApi<B> + Send + Sync,
+{
+	// empty
+}
+
 /// Allows all gossip messages to get through.
 struct AllowAll<Hash> {
 	topic: Hash,
 }
 
-impl<Block> GossipValidator<Block> for AllowAll<Block::Hash>
+impl<B> GossipValidator<B> for AllowAll<B::Hash>
 where
-	Block: BlockT,
+	B: Block,
 {
 	fn validate(
 		&self,
-		_context: &mut dyn GossipValidatorContext<Block>,
+		_context: &mut dyn GossipValidatorContext<B>,
 		_sender: &sc_network::PeerId,
 		_data: &[u8],
-	) -> GossipValidationResult<Block::Hash> {
+	) -> GossipValidationResult<B::Hash> {
 		GossipValidationResult::ProcessAndKeep(self.topic)
 	}
 }
 
-pub async fn start_beefy_gadget<Block, Pair, Backend, Client, Network, SyncOracle>(
-	client: Arc<Client>,
+/// Start the BEEFY gadget.
+///
+/// This is a thin shim around running and awaiting a BEEFY worker. The [`Client`]
+/// convenience trait is not used here on purpose. We don't want to leak it into the
+/// public interface of the BEEFY gadget.
+pub async fn start_beefy_gadget<B, P, BE, C, N, SO>(
```

Comment:
I'm not a huge fan of this one-letter generic parameters, it's so easy to forget what they are supposed to represent. Personally I like the (java?) `TBlock`, `TPair`, etc convention where `T` makes it clear it's a generic parameter, but the rest of the name gives you some clue what it actually is.

Reply:
This is the first case, we really differ in preferences, I guess. For me a generic parameter is, well generic, i.e. meaningless. The meaning or semantic is induced be the where clause. I also don't like to duplicate information which is easily provided by common tools like rust-analyzer.

Reply:
Question is, how we want to deal with those kind of preferences, both as a team and as a company. Establish and enforce a style guide or accept different preferences. As always, both options have pros and cons.

Reply:
We don't have a formal framework how to deal with that, but I'm not strong enough on my preference to keep defending it :) I can work in both, and prefer consistency over any specific choice. Since it's mainly you working on that repo, I think you should be the one making the decision.

Reply:
Appreciate that!